### ISSUE TITLE
Is the performance right for lmdb training(imagenet-resnet) ?

### ISSUE THREAD

**Issue by tjulyz** [ppwwyyxx/tensorpack#672]
The configuration : gpu-4x1080ti, cpu-xeon 32G, HDD. cuda-8.0, cudnn-6.0
Training resnet50 on imagenet by the original code example: 0.4-0.5 iters
Due to the low speed, I use the code 
```
from tensorpack.dataflow import *
class BinaryILSVRC12(dataset.ILSVRC12Files):
    def get_data(self):
        for fname, label in super(BinaryILSVRC12, self).get_data():
            with open(fname, 'rb') as f:
                jpeg = f.read()
            jpeg = np.asarray(bytearray(jpeg), dtype='uint8')
            yield [jpeg, label]
ds0 = BinaryILSVRC12()
ds1 = PrefetchDataZMQ(ds0, nr_proc=1)
dftools.dump_dataflow_to_lmdb(ds1, '/path/to/ILSVRC-train.lmdb')
```
according to your tutorials to generate lmdb.
And replace the original code in imagenet_utils.py
```
if isTrain:
        ds = dataset.ILSVRC12(datadir, name, shuffle=True)
        ds = AugmentImageComponent(ds, augmentors, copy=False)
        if parallel < 16:
            logger.warn("DataFlow may become the bottleneck when too few processes are used.")
        ds = PrefetchDataZMQ(ds, parallel)
        ds = BatchData(ds, batch_size, remainder=False)
```
by
```
ds = LMDBData('/path/to/ILSVRC-train.lmdb', shuffle=False)
ds = LocallyShuffleData(ds, 50000)
ds = PrefetchData(ds, 5000, 1)
ds = LMDBDataPoint(ds)
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
ds = AugmentImageComponent(ds, augmentors)
ds = PrefetchDataZMQ(ds, parallel)
ds = BatchData(ds, 256)
```
for reading training data from the generated lmdb without changing the reading manner of validation.
Then I get the training speed : 2.0-2.2 iters
What makes me confused is the performance. Using the original code, the validation errors drop significantly. For instance, after training 15 epochs, the top-1/5 validation errors drop to 0.43/0.19(training is 0.48/0.24). But with the changed code, the error rates drop too slowly (top-1/5 validation errors are 0.89/0.79 after 15 epochs, training is 0.36/0.14). 
So is the performance right for the changed code? 

Thanks a lot!

**Comment by ppwwyyxx** [comment]
For speed, read tutorial: http://tensorpack.readthedocs.io/en/latest/tutorial/performance-tuning.html
For accuracy: I don't know what happened -- any line of code that's written wrong could cause it. If you can post more information following the issue template maybe I'll be able to tell more.

**Comment by tjulyz** [comment]
I didn't change other codes but only the codes for reading training data from LMDB file. The codes are copyed from your tutorials and shown as above.

**Comment by ppwwyyxx** [comment]
> The codes are copyed from your tutorials and shown as above.

Then you're using a different batch size compared to the original code.

**Comment by tjulyz** [comment]
Sorry about that. In my running codes,        ds = BatchData(ds, batch_size)

**Comment by ppwwyyxx** [comment]
It's better to paste the changes (in the form of diff, for example) you made instead of describing them. As you can see from the comment above it's not easy to describe them accurately.

In addition to that there are a lot of other issues that may matter. What's the version of tensorflow? How do you install tensorpack? For the two runs you mentioned originally, are they on the same system with consistent software stack?

**Comment by tjulyz** [comment]
1. ![image](https://user-images.githubusercontent.com/18508335/36660813-65448cce-1b14-11e8-82c9-5094e83df48e.png)

2. ![image](https://user-images.githubusercontent.com/18508335/36660741-31be2fb8-1b14-11e8-87a6-f7d70d3a594b.png)

I run the two on the same computer with same environment: TF1.40, Python2.7, and install the tensorpack use: pip install -U git+https://github.com/ppwwyyxx/tensorpack.git.

**Comment by ppwwyyxx** [comment]
I don't know what went on. This is not an issue I can reproduce, but the code looks fine. Maybe check the lmdb file, something may went wrong when you generate it. For example if it's smaller than 140G it's incomplete.

**Comment by tjulyz** [comment]
What the manner do you choose for training your resnet models? With LMDB file or reading from list randomly? 
Besides, is the performance reported in your README of Resnet evaluated on validation dataset without pre-activation?
Thanks a lot!

**Comment by ppwwyyxx** [comment]
I read raw jpegs. But reading from LMDB will have similar performance.

The numbers in the first table does not use preact.

**Comment by tjulyz** [comment]
I have checked the generated LMDB. It is about 149.7GB.
I will change the environment and try it again.
Thank you for your share and kind reply!
Thanks a lot!
